First completed draft of ID Generation post.

nathankleyn opened the pull request:
Also added author image, and post image and thumbnails. Added `.gitkeep`
file to the `_drafts` folder to stop it from disappearing every time we
publish our only draft.

nathankleyn committed:
First completed draft of ID Generation post.

Also added author image, and post image and thumbnails. Added `.gitkeep`
file to the `_drafts` folder to stop it from disappearing every time we
publish our only draft.

albertpastrana commented on _drafts/id-generation.md:
```diff
@@ -0,0 +1,104 @@
+
+**TL;DR:** We are open-sourcing our distributed _k_-ordered ID generation project called "Icicle", which generates IDs using Lua scripting within distributed Redis hosts. You can [find more information in the GitHub repository](https://github.com/intenthq/icicle)!
+
+At Intent HQ, we generate a lot of data. Part of the task of capturing this data is the package it up into a common format and store it in a way it can be retrieved. Commonly, labelling this data in a way that facilitates retrival is a task handled by your database. However, in a big-data environment, we need a way to label these pieces of data with an ID in a distributed and scalable way indepen ...
```
`...capturing the data is the package it up...` -> `...capturing the data is TO package it up...`

`retrival` -> `retrieval`

albertpastrana commented on _drafts/id-generation.md:
```diff
@@ -0,0 +1,104 @@
+
+## What is an ID?
+
+An ID can be many things, but surely its most important definition is that of "uniqueness". Of all the properties, this is the only one that truly matters. If you label a piece of data with an ID, you want to be sure you don't later have abiguities when looking up that data again by the ID you had.
```
`abiguities` -> `ambiguities`

nathankleyn committed:
Run post through spell checker, fix review comments.

philmes commented on _drafts/id-generation.md:
```diff
@@ -0,0 +1,104 @@
+
+* _Meaningful:_ They may embed certain information about the point at which the ID was generated, for example the time (however inexact).
+
+* _Distributed:_ They may be generated by a machine without consulating any other machine (ie. they must be truly distributed).
```
consulating -> consulting

albertpastrana commented on _drafts/id-generation.md:
```diff
@@ -0,0 +1,104 @@
+
+Keeping time is actually a really hard problem to solve. At the core of the issue is the fact that computers suffer from "clock drift", where the clock is constantly skewing forwards or backwards away from the actual time. Perhaps even worse, [every computer skews at a different rate](https://en.wikipedia.org/wiki/Clock_skew).
+
+Thus, the concept of a "time oracle" is born. Time oracles dedicated servers normally equipped with an atomic clock or GPS device whose sole purpose is for keeping time in as precise a manner as possible. Time oracle servers are used frequently in environments where super accurate time is important (such as trading platforms and financial applications), but they are prohibitively expensive and no ...
```
More than "those unable to manage their own physical server" I would probably write "those we don't manage their own physical server".

nathankleyn committed:
Optimise images for size, fix review comments, add some extra points.

albertpastrana commented on _drafts/id-generation.md:
```diff
@@ -0,0 +1,104 @@
+
+* D is the sequence, 12 bits in total.
+
+We have represented the time as a 41-bit integer by creating a custom epoch: this simply involves picking an arbitrary point in time, calling it "0", and starting to count the milliseconds from there.
```
Maybe we should say here how many time do we have? (~69y)

albertpastrana commented on _drafts/id-generation.md:
```diff
@@ -0,0 +1,104 @@
+The final piece of the puzzle is in how we use Redis to do this.
+
+Redis lacks the ability to run commands in a transaction based on logic, so we chose to implement the logic in Lua and pass it to Redis using its scripting functionality. At the time of coding, the bitops library provided by Redis didn't allow us to do what we needed to do to compose the 64-bit ID in pure Lua, so we currently call to Redis with the Lua script to get the components which make up t ...
+
```
I would add here pseudo-code showing how we do it.

nathankleyn committed:
Updated blog post and styles.

Consulted with Mark to darken the text styles after feedback on Hacker
News and Reddit about contrast being an issue.

Added images, and entire new section to post about performance and
compaction of data representation.

nathankleyn committed:
Change k-ordered to k-sorted, more idiomatic term.

nathankleyn committed:
Moved ID generation post into posts folder, added links to repo.
